chalicelib/matrix_handler.py

```diff
@@ -0,0 +1,141 @@
+import json
+import os
+import tempfile
+import boto3
+import hca
+import loompy
+
+from abc import ABC, abstractmethod
+from botocore.exceptions import ClientError
+from chalicelib.constants import MERGED_MTX_BUCKET_NAME, MERGED_REQUEST_STATUS_BUCKET_NAME, JSON_EXTENSION
+from chalicelib.request_handler import RequestHandler, RequestStatus
+from chalicelib import logger
+
+
+class MatrixHandler(ABC):
+    """
+    A generic matrix handler for matrices concatenation
+    """
+    def __init__(self, extension):
+        self._extension = extension
+        self.hca_client = hca.dss.DSSClient()
+
+    def _filter_mtx(self, bundle_uuids):
+        """
+        Filter only matrix files from a list of DSS bundles.
+        :param bundle_uuids: A list of bundle uuids.
+        :return: A list of matrices uuids.
+        """
+        mtx_uuids = []
+
+        # Iterate uuids to query DSS for bundle manifest of each
+        for uuid in bundle_uuids:
+            bundle_manifest = self.hca_client.get_bundle(replica="aws", uuid=uuid)
+
+            # Gather up uuids of all the matrix files we are going to merge
+            for file in bundle_manifest["bundle"]["files"]:
+                if file["name"].endswith(self._extension):
+                    mtx_uuids.append(file["uuid"])
+
+        return mtx_uuids
+
+    def _download_mtx(self, bundle_uuids):
+        """
+        Filter for the matrix files within bundles, and download them locally
+        TODO: Is there a way to directly connect to a s3 file instead of downloading it locally?
+
+        :param bundle_uuids: A list of bundle uuids
+        :return: A list of downloaded local matrix files paths
+        """
+        logger.info("Downloading matrices from bundles: %s.", str(bundle_uuids))
+
+        # Filter uuids of matrix files within each bundle
+        mtx_uuids = self._filter_mtx(bundle_uuids)
+
+        # Create a temp directory for storing s3 matrix files
+        temp_dir = tempfile.mkdtemp()
+
+        local_mtx_paths = []
+
+        # Download matrices from cloud to a local temp directory
+        for uuid in mtx_uuids:
+            path = os.path.join(temp_dir, uuid + self._extension)
+            with open(path, "wb") as mtx:
+                mtx.write(self.hca_client.get_file(uuid=uuid, replica="aws"))
```

Comment:
This seems sensible, but the `download` function in the DCP CLI does something quite a bit more complex when it tries to download files from a bundle. I'm guessing it's trying to, among other things, avoid reading the entire file into memory before writing it, but take a look: https://github.com/HumanCellAtlas/dcp-cli/blob/master/hca/dss/__init__.py#L35